modules/apps/feature-flag/feature-flag-web/src/test/java/com/liferay/feature/flag/web/internal/feature/flag/FeatureFlagsBagFactoryTest.java

```diff
@@ -0,0 +1,181 @@
+import com.liferay.portal.kernel.feature.flag.FeatureFlag;
+import com.liferay.portal.kernel.feature.flag.constants.FeatureFlagConstants;
+import com.liferay.portal.kernel.language.Language;
+import com.liferay.portal.kernel.model.CompanyConstants;
+import com.liferay.portal.test.log.LogCapture;
+import com.liferay.portal.test.log.LogEntry;
+import com.liferay.portal.test.log.LoggerTestUtil;
+import com.liferay.portal.test.rule.LiferayUnitTestRule;
+import com.liferay.portal.util.PropsImpl;
+import com.liferay.portal.util.PropsUtil;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.logging.Level;
+
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.Test;
+
+import org.mockito.Mockito;
+
+/**
+ * @author Thiago Buarque
+ */
+public class FeatureFlagsBagFactoryTest {
+
+	@ClassRule
+	@Rule
+	public static final LiferayUnitTestRule liferayUnitTestRule =
+		LiferayUnitTestRule.INSTANCE;
+
+	@Before
+	public void setUp() {
+		com.liferay.portal.kernel.util.PropsUtil.setProps(new PropsImpl());
+	}
+
+	@Test
+	public void testFeatureFlagSelfDependency() {
+		long companyId = 1L;
+
+		String key = "LPS-1010";
+
+		PropsUtil.set(FeatureFlagConstants.getKey(key), "true");
+		PropsUtil.set(FeatureFlagConstants.getKey(key, "dependencies"), key);
+
+		try (LogCapture logCapture = LoggerTestUtil.configureJDKLogger(
+				FeatureFlagsBagFactory.class.getName(), Level.SEVERE)) {
+
+			_featureFlagsBagFactory.create(
+				companyId, _featureFlagPreferencesManager, _language);
+
+			List<LogEntry> logEntries = logCapture.getLogEntries();
+
+			LogEntry logEntry = logEntries.get(0);
+
+			Assert.assertEquals(
+				String.format(
+					"A feature flag cannot depend on itself: %s", key),
+				logEntry.getMessage());
+		}
+	}
+
+	@Test
+	public void testNonsystemFeatureFlags() {
+		long companyId = 1L;
+
+		String key1 = "LPS-123";
+
+		PropsUtil.set(FeatureFlagConstants.getKey(key1), "true");
+		PropsUtil.set(FeatureFlagConstants.getKey(key1, "system"), "true");
+
+		String key2 = "LPS-456";
+
+		PropsUtil.set(FeatureFlagConstants.getKey(key2), "true");
+		PropsUtil.set(FeatureFlagConstants.getKey(key2, "system"), "false");
+
+		String key3 = "LPS-789";
+
+		PropsUtil.set(FeatureFlagConstants.getKey(key3), "true");
+
+		FeatureFlagsBag featureFlagsBag = _featureFlagsBagFactory.create(
+			companyId, _featureFlagPreferencesManager, _language);
+
+		List<String> keys = Arrays.asList(key1, key2, key3);
+
+		List<FeatureFlag> featureFlags = featureFlagsBag.getFeatureFlags(
+			featureFlag -> keys.contains(featureFlag.getKey()));
+
+		Assert.assertTrue(featureFlags.size() == 2);
+
+		List<String> nonsystemKeys = Arrays.asList(key2, key3);
+
+		FeatureFlag featureFlag = featureFlags.get(0);
+
+		Assert.assertTrue(nonsystemKeys.contains(featureFlag.getKey()));
+
+		featureFlag = featureFlags.get(1);
+
+		Assert.assertTrue(nonsystemKeys.contains(featureFlag.getKey()));
```

Comment:
It's probably worth checking that trying to lookup the value of a non-system feature flag with a the companyId=0 logs the info message.

Reply:
The test you mentioned belongs to FeatureFlagsBagTest, just added it there